File: Android/app/src/main/java/org/p9/spear/component/Port.kt
package org.p9.spear.component

import android.net.VpnService
import android.util.Log
import org.p9.spear.entity.Packet
import java.lang.Exception
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.SocketAddress
import java.net.StandardSocketOptions
import java.nio.ByteBuffer
import java.nio.channels.DatagramChannel

fun getSocketAddressByEndpoint(endpoint: String): SocketAddress {
    val ipPort = endpoint.split(":")
    val ipAddress = InetAddress.getByName(ipPort[0])
    val port = Integer.parseInt(ipPort[1])
    return InetSocketAddress(ipAddress, port)
}

abstract class IPort (protected val vpnService: VpnService) {

    protected var connected: Boolean? = false

    abstract fun connect(): Boolean

    abstract fun stop(): Boolean

    abstract fun send(packet: Packet): Boolean

    var onReceive: ((Packet) -> Boolean)? = null
}

class TestPort(vpn: VpnService, private val endpoint: String) : IPort(vpn) {

    private var tunnel: DatagramChannel? = null
    private var thread: Thread? = null
    private val packet = Packet(
        ByteBuffer.allocate(UShort.MAX_VALUE.toInt()))

    override fun connect(): Boolean {
        if (connected != false) {
            return false
        }
        connected = null
        try {
            val t = DatagramChannel.open()
            if (!vpnService.protect(t.socket())) {
                throw IllegalStateException("Cannot protect the tunnel")
            }
            t.configureBlocking(true)
            t.setOption(StandardSocketOptions.SO_SNDBUF, 16 * 1024 * 1024)
            t.setOption(StandardSocketOptions.SO_RCVBUF, 16 * 1024 * 1024)
            t.connect(getSocketAddressByEndpoint(endpoint))
            tunnel = t
            thread = Thread {
                loop()
            }
            thread?.start()
        }
        catch (ex: Exception) {
            Log.e(javaClass.name, "connect exception", ex)
            connected = false
            return false
        }

        connected = true
        return true
    }

    override fun stop(): Boolean {
        if (connected != true) {
            return false
        }
        connected = null
        thread?.interrupt()
        thread = null
        tunnel = null
        connected = false
        return true
    }

    override fun send(packet: Packet): Boolean {
        if (connected != true || tunnel == null) {
            return false
        }
        tunnel?.write(packet.buffer)
        return true
    }

    private fun loop() {
        while(connected == true && tunnel != null) {
            packet.len = tunnel?.read(packet.buffer) ?: 0
            if (packet.len > 0) {
                if (onReceive?.invoke(packet) == false) {
                    break
                }
                packet.buffer.clear()
            } else {
                Log.e(javaClass.name, "loop but no data!!!!")
            }
        }
    }
}

File: iOS/TunnelProvider/Entity/Port.swift
//
//  Port.swift
//  TunnelProvider
//
//  Created by dexterdreeeam on 2024/1/3.
//

import Foundation
import NetworkExtension

protocol IPortProtocol {
    
    func connect() -> Bool
    
    func stop() -> Bool
    
    func send(packet: Packet) -> Bool
    
    var onReceive: ((Packet) -> Bool)? { get set }
}

class IPort: IPortProtocol {
    
    internal var connected: Bool? = false
    
    func connect() -> Bool {
        fatalError("Not Implement")
    }
    
    func stop() -> Bool {
        fatalError("Not Implement")
    }
    
    func send(packet: Packet) -> Bool {
        fatalError("Not Implement")
    }
    
    var onReceive: ((Packet) -> Bool)? {
        get {
            fatalError("Not Implement")
        }
        set {
            fatalError("Not Implement")
        }
    }
}

class TestPort: IPort {
    
    private var endpoint: String
    private var thread: Thread?
    
    init(endpoint: String) {
        self.endpoint = endpoint
    }
    
    override func connect() -> Bool {
        guard self.connected == false else {
            return false
        }
        self.connected = nil
        do {
            // todo udp
        
            self.thread = Thread {
                self.loop()
            }
            self.thread!.start()
        } catch {
            // error
            connected = false
            return false
        }
        connected = true
        return true
    }
    
    override func stop() -> Bool {
        guard self.connected == true else {
            return false
        }
        self.connected = nil
        self.thread?.cancel()
        self.thread = nil
        self.connected = false
        return true
    }
    
    override func send(packet: Packet) -> Bool {
        guard self.connected == true else {
            return false
        }
        // todo
        return true
    }
    
    private func loop() {
        
    }
}
